tests/test_models.py

```diff
@@ -284,6 +291,164 @@ async def test_device_json(aresponses: ResponsesMockServer) -> None:
+            host="example.com",
+            source_type="tcp",
+            serial_number=serial_number,
+            _socket_mock=socket_mock,
+        )
+
+        def send_side_effect(data: bytes) -> int:
+            assert data == tcp.create_information_request(serial_number)
+            asynctest.set_read_ready(socket_mock, self.loop)
+            return len(data)
+
+        # Validate message start and end markers
+
+        socket_mock.send.side_effect = send_side_effect
+        socket_mock.recv.side_effect = lambda _: b"broken data"
+
+        with pytest.raises(OmnikInverterPacketInvalidError) as excinfo:
+            assert await client.inverter()
+
+        assert str(excinfo.value) == "Invalid start byte"
+
+        socket_mock.recv.side_effect = lambda _: bytearray([tcp.MESSAGE_START, 0])
+
+        with pytest.raises(OmnikInverterPacketInvalidError) as excinfo:
+            assert await client.inverter()
+
+        assert str(excinfo.value) == "Invalid end byte"
+
+        # Test checksum calculation through false checksum
+
+        test_message = b"foo"
+        false_checksum = 0xFE
+        checksum = sum(test_message) & 0xFF
+
+        socket_mock.recv.side_effect = lambda _: bytearray(
+            [tcp.MESSAGE_START] + list(test_message) + [false_checksum, tcp.MESSAGE_END]
+        )
+
+        with pytest.raises(OmnikInverterPacketInvalidError) as excinfo:
+            assert await client.inverter()
+
+        assert (
+            str(excinfo.value)
+            == f"Checksum mismatch (calculated `{checksum}` got `{false_checksum}`)"
+        )
+
+        # Test receiving valid data, but with mismatching serial number
+
+        def recv_side_effect(_max_bytes: int) -> bytes:
+            return load_fixture_bytes("tcp_reply.data")
+
+        socket_mock.recv.side_effect = recv_side_effect
+
+        with pytest.raises(OmnikInverterPacketInvalidError) as excinfo:
+            assert await client.inverter()
+
+        assert str(excinfo.value) == "Serial number mismatch in reply"
+
+    async def test_connection_broken(self) -> None:
+        """Test on broken connection - TCP source."""
+        serial_number = 1
+        socket_mock = asynctest.SocketMock()
+        socket_mock.type = socket.SOCK_STREAM
+
+        client = OmnikInverter(
+            host="example.com",
+            source_type="tcp",
+            serial_number=serial_number,
+            _socket_mock=socket_mock,
+        )
+
+        def send_side_effect(data: bytes) -> int:
+            assert data == tcp.create_information_request(serial_number)
+            asynctest.set_read_ready(socket_mock, self.loop)
+            return len(data)
+
+        socket_mock.send.side_effect = send_side_effect
+        socket_mock.recv.side_effect = Exception("Connection broken")
+
+        with pytest.raises(Exception) as excinfo:
+            assert await client.inverter()
+
+        assert str(excinfo.value) == "Connection broken"
```

Comment:
Not entirely sure what's going on here. `pylint` wants me to test:

https://github.com/klaasnicolaas/python-omnikinverter/blob/935ab484154dc1ddcf5f96892cf152c4f11bb988/omnikinverter/omnikinverter.py#L151-L154

Yet I can't check for that specific exception here, even though it is clearly visible in the logs. Instead, I have to check for the much less specific `Exception("Connection broken")` here that is raised by `asynctest` for us.

Reply:
I found what is going on, and the details are in comparing `raise ... from ...` re-wrapping (with `The above exception was the direct cause of the following exception:`) working in `test_connection_failed()`, but not in this `test_connection_broken(self)` which shows `The above exception was the direct cause of the following exception:`.

Turns out the exception is (captured and re-) raised in the `finally:` block through the `.wait_closed()` call, which doesn't invoke our `raise OmnikInverterConnectionError(...) from ...`. This happens while handling the `OSError` exception as stated by the error message.